### PR TITLE
ENH (WIP): Adds Poisson binomial distribution support (See #6000)

### LINKED ISSUE #6000: ENH: stats: add Poisson Binomial Distribution
I'd like to add support for the Poisson Binomial Distribution: https://en.wikipedia.org/wiki/Poisson_binomial_distribution

into the scipy.stats module. It could be placed under "scipy.stats.pbinom". Is there a reason why support for this doesn't already exist and if not, would it be a welcome addition to the scipy repo?

If so, I'll go ahead and implement it and submit a PR.

### PR DESCRIPTION
**Pull request by maxi-marufo**
#### Reference issue
Closes #6000

#### What does this implement/fix?
This PR adds the Poisson binomial distribution to the discrete distributions.
As the computing of its _pmf_ of is not straightforward, this implementation uses the derivation for an exact formula with a closed-form expression that uses the discrete Fourier transform of the characteristic function of the distribution, as proposed in [Hong2013](https://dx.doi.org/10.1016/j.csda.2012.10.006). The original Python implementation of the code was proposed by [Mika Straka](https://github.com/tsakim/poibin).

For details see [Poisson binomial distribution in Wikipedia](https://en.wikipedia.org/wiki/Poisson_binomial_distribution).

#### Additional information

Tests, benchmarks and docs have also been added.

**This PR still needs more work**:

1. When running `make html-scipyorg` I get an error (`TypeError: __call__() missing 1 required positional argument: 'probs'`) which I assume is generated when calling `_construct_doc`. Help for solving this issue is welcomed.

2. Some tests (`test_discrete_basic[pbinom-arg14-True]` and `test_moments[pbinom-arg14]`) are failing. This also is related to the addition of `probs` as a new argument of the distribution.

Also, several magic methods (`__init__, __new__, __call__`) and normal methods (`_updated_ctor_param, _argcheck`) were overridden to implement this class. Additional work could be added to make this implementation more consistent with the rest of the classes (I've tried to use `rv_samples` instead of `rv_discrete`, but without good results)

Thanks to [Mika Straka](https://github.com/tsakim/poibin) and [YIl Hong](https://dx.doi.org/10.1016/j.csda.2012.10.006) for their original contributions, and [Hamel Ajay Kothari](https://github.com/hkothari) for his [initial work](https://github.com/scipy/scipy/issues/6000).

Funny fact: PR #12000 linked to issue #6000

### CODE REVIEW

**Comment by pvanmulbregt** [comment]
@maxi-marufo One the one hand this is really interesting.  On the other hand, I really don't like what you had to do to make it work.
The fact that you had to override the magic methods `__init__, __new__, __call__` and `_updated_ctor_param` is a signal that there is a mismatch here between what you want to do and the infrastructure available to support it.

IIUC, there are two reasons for the approach taken here.
1. `probs=[0.4, 0.2]` will be broadcast, rather than treated as a single argument to the constructor.
2. The computations for the `pmf` can be quite time consuming, and having to do them each time is undesirable.

I'll give here some ideas on how to work around these issues within the current framework. [Another approach would be to create a generic framework that supported storing shape-specific values in the class.]

For the first item, I can suggest creating a `ProbabilityList` class, and then one could pass in `probs=ProbabilityList(0.4, 0.2, ..)`, or `probs=ProbabilityList([0.4, 0.2, ...])` and that would not be broadcast.

The second item is more problematic.  The standard infrastructure for distributions built on `rv_continuous`, `rv_discrete` no longer permits setting shape-specific member variables in the class. The reason is that there is only a single instance of the class which is shared by all users, and allowing all users of the class to shape-specific values breaks thread safety. And frozen distributions don't permit it either, as a frozen distribution is really just a container for (shape parameters, pointer to the regular distribution) and doesn't store results of computation(s).
[I ran into a similar problem with the truncated normal distribution `truncnorm`, which was storing the normalization factor in the class. gh-9900 removed that, and it took a couple of PRs gh-10104, gh-11393 to recover speed.]

The impact of not being able to store the `_pmf_list` and `_cdf_list` in the `pbinom_gen` class depends on the use cases.  
 - If a single `pmf` is being requested for each distinct collection of `probs`, then it doesn't matter.
 - If there is a single call to `pmf(x, ...)` with a long list of `x` values, it also doesn't matter.  
- If there are many calls, for the same fixed `probs`, to `pmf(x, ...)` with a single `x`, then that will take longer.

Same is true for `cdf` and the other member functions.  Having a custom `_rvs` method is often needed in this type of situation, as you did here. 

Can you add any information on the use cases that are important for this distribution?

**Comment by maxi-marufo** [comment]
> @maxi-marufo One the one hand this is really interesting. On the other hand, I really don't like what you had to do to make it work.
> The fact that you had to override the magic methods `__init__, __new__, __call__` and `_updated_ctor_param` is a signal that there is a mismatch here between what you want to do and the infrastructure available to support it.
> 
> IIUC, there are two reasons for the approach taken here.
> 
> 1. `probs=[0.4, 0.2]` will be broadcast, rather than treated as a single argument to the constructor.
> 2. The computations for the `pmf` can be quite time consuming, and having to do them each time is undesirable.
> 
> I'll give here some ideas on how to work around these issues within the current framework. [Another approach would be to create a generic framework that supported storing shape-specific values in the class.]
> 
> For the first item, I can suggest creating a `ProbabilityList` class, and then one could pass in `probs=ProbabilityList(0.4, 0.2, ..)`, or `probs=ProbabilityList([0.4, 0.2, ...])` and that would not be broadcast.
> 
> The second item is more problematic. The standard infrastructure for distributions built on `rv_continuous`, `rv_discrete` no longer permits setting shape-specific member variables in the class. The reason is that there is only a single instance of the class which is shared by all users, and allowing all users of the class to shape-specific values breaks thread safety. And frozen distributions don't permit it either, as a frozen distribution is really just a container for (shape parameters, pointer to the regular distribution) and doesn't store results of computation(s).
> [I ran into a similar problem with the truncated normal distribution `truncnorm`, which was storing the normalization factor in the class. [gh-9900](https://github.com/scipy/scipy/pull/9900) removed that, and it took a couple of PRs [gh-10104](https://github.com/scipy/scipy/pull/10104), [gh-11393](https://github.com/scipy/scipy/pull/11393) to recover speed.]
> 
> The impact of not being able to store the `_pmf_list` and `_cdf_list` in the `pbinom_gen` class depends on the use cases.
> 
> * If a single `pmf` is being requested for each distinct collection of `probs`, then it doesn't matter.
> * If there is a single call to `pmf(x, ...)` with a long list of `x` values, it also doesn't matter.
> * If there are many calls, for the same fixed `probs`, to `pmf(x, ...)` with a single `x`, then that will take longer.
> 
> Same is true for `cdf` and the other member functions. Having a custom `_rvs` method is often needed in this type of situation, as you did here.
> 
> Can you add any information on the use cases that are important for this distribution?

@pvanmulbregt Thank you very much for your comments, I appreciate your detailed explanation of the problems. I agree that overriding those magic methods shows some incompatibility, but anyway I wanted to push the commits to see if someone with more experience with the framework suggest a different approach, as I am just getting used to the code.
I've just pushed some new changes: I've overriden `_get_support` (which wasn't there before) and also removed the easiest member variables. I left the `probs`, `_pmf_list` and `_cdf_list` to see if anyone has other suggestions.
I think the most usual use case would be to compute some range of the distribution, or all of it, i.e.: the second use case you mentioned. I also think the third use case would be naturally avoided and instead the second one would be used when some sparse `x` are used, specially in these more computational expensive distributions. Maybe a notice in the documentation regarding this issue could be added? (e.g: calling `pmf(int(...))` and pmf([...])` is almost equally computational expensive, so try to avoid calling it many times and instead use one call with all the `x` values, or something similar?).
Anyway, `_pmf_list` and `_cdf_list` could be easily removed (with this last consideration), but I would still like to read more suggestions about the use of the `probs` argument. Although maybe using the `ProbabilityList` you mentioned is a good solution.

**Comment by maxi-marufo** [comment]
@pvanmulbregt I've been trying to follow your suggestions to make this work. I don't want to push the code because I now I get other errors (which I'll describe later), but these are the genreal changes:
- I created a `ProbabilityList`, which I use for the `probs` argument. This class has only the `__init__` method which converts the input `list` to `np.array` and stores it as a member variable.
- I deleted all the magic methods, `_updated_ctor_param` and `freeze_pbinom` from `pbinom_gen`
- All the methods from `pbinom_gen` now have `probs` as argument.
- I compute `_pmf_list` and `_cdf_list` every time I need it.

But now I'm running into these 2 problems:
- I think `_distr_params.py` is not prepared to use a list as argument. Because of that, I'm having problems with `test_discrete_basic.py`. (I already have this problem with the current code in the PR)
- I get either `IndexError: cannot do a non-empty take from an empty axes.` or `ValueError: operands could not be broadcast together with shapes` (depending on the test) when trying to execute `return [np.extract(cond, arr1 * expand_arr) for arr1 in newargs]` in `test_distributions.py::TestPBinom`.

When you have some time, could you please provide some example code on how to approach your suggestions?
Thanks!

**Comment by pvanmulbregt** [comment]
@maxi-marufo Thanks for the investigation.  Apparently this is not going to be an obvious one to crack. I'll get to think about it some more in a few days.

File: THANKS.txt [hunk 1]
```diff
@@ -234,6 +234,8 @@ James Wright for simple documentation fixes
 Sam Wallan for scipy.linalg.lapack enhancements
 Richard Weiss for a bug fix in scipy.optimize._differentialevolution.py.
 Luigi F. Cruz for adding time/frequency domain option to scipy.signal.resample.
+Yili Hong, Mika Straka, Hamel Ajay Kothari and Maxi Marufo for adding Poisson
```
Comment (rlucas7) [inline THANKS.txt:237]: Can these lines be removed from the PR? 
See [thread](https://mail.python.org/pipermail/scipy-dev/2020-August/024331.html) for discussion (changes to this file causes merge conflicts adding effort to contributors and maintainers). 
For now, I've opened a [wiki](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.6.0*)
If/once the PR is merged, we add thanks to the wiki. Similar process to what we followed in 1.5.0 release.

**Comment by mdhaber** [comment]
It doesn't look like this distribution will fit in the existing infrastructure, and this PR has developed many merge conflicts since it was last developed, so I'll go ahead and close it. gh-6000 is still open to track the request to add this distribution, and gh-15928 tracks the need for theb infrastructure to support distributions like this. Thanks for your work here! If you don't mind, please keep the branch so we can use it as a starting point when the infrastructure supports it. Perhaps we can even bring back this PR itself - it would be fun to close gh-6000 with gh-12000.

**Pull request closed without merging**